src/aics-image-viewer/components/TfEditor/index.tsx

```diff
@@ -82,669 +67,313 @@ const TF_GENERATORS: Record<string, (histogram: Histogram) => Lut> = {
-
-    // Scale data range to 8bit
-    this.dataScale = d3.scaleLinear();
-
-    // Keep track of control points interaction
-    this.dragged = null;
-    this.selected = null;
-    this.last_color = TFEDITOR_DEFAULT_COLOR;
-  }
+const clamp = (value: number, min: number, max: number): number => Math.min(Math.max(value, min), max);
 
-  private initializeElements(): void {
-    let extent = [0, 255];
-    if (this.fitToData && this.props.volumeData && this.props.volumeData.length > 0) {
-      extent = d3.extent(this.props.volumeData) as Pair;
-    }
-    this.xScale.rangeRound([0, this.width]).domain(extent);
-    this.yScale.domain([0, 1]).range([this.height, 0]);
-    this.binScale.domain([1, 10]).range([this.height, 0]).base(2).clamp(true);
-    this.bins.domain(this.xScale.domain() as Pair).thresholds(this.xScale.ticks(this.numberBins));
-    if (this.props.controlPoints.length === 0) {
-      let newControlPoints = [
-        {
-          x: extent[0],
-          opacity: 0,
-          color: TFEDITOR_DEFAULT_COLOR,
-        },
-        {
-          x: extent[1],
-          opacity: 1,
-          color: TFEDITOR_DEFAULT_COLOR,
-        },
-      ];
-      this.props.updateChannelLutControlPoints(newControlPoints);
-    }
-    this.selected = this.props.controlPoints[0];
-    this.area
-      .x((d) => this.xScale(d.x))
-      .y0((d) => this.yScale(d.opacity))
-      .y1(this.height)
-      .curve(d3.curveLinear);
+const TfEditor: React.FC<TfEditorProps> = (props) => {
+  const innerWidth = props.width - TFEDITOR_MARGINS.left - TFEDITOR_MARGINS.right;
+  const innerHeight = props.height - TFEDITOR_MARGINS.top - TFEDITOR_MARGINS.bottom;
 
-    this.canvasScale.range([0, 1]);
-    this.dataScale.domain(extent).range([0, 255]);
+  const [selectedPointIdx, setSelectedPointIdx] = useState<number | null>(null);
+  const [draggedPointIdx, _setDraggedPointIdx] = useState<number | null>(null);
 
-    // Canvas element selector to output the result
-    this.canvasSelector = this.canvasSelector || "#canvas-" + this.id;
-  }
+  // these bits of state need their freshest, most up-to-date values available in mouse event handlers. make refs!
+  const [controlPointsRef, setControlPoints] = useRefWithSetter(props.updateLutControlPoints, props.controlPoints);
+  const [draggedPointIdxRef, setDraggedPointIdx] = useRefWithSetter(_setDraggedPointIdx, draggedPointIdx);
 
-  // Get the 2D canvas context where the TF will be drawn
-  private canvasContext(): CanvasRenderingContext2D | null {
-    let canvas_element = this.canvas.current;
-    if (canvas_element) {
-      return canvas_element.getContext("2d");
-    }
-    return canvas_element;
-  }
+  // Either `null` when the control panel is closed, or an x offset into the plot to position the color picker.
+  // Positive: offset right from the left edge of the plot; negative: offset left from the right edge of the plot.
+  const [colorPickerPosition, setColorPickerPosition] = useState<number | null>(null);
```

Comment:
does this also need an "above"/"below" designation too?  For when the tf editor is near the bottom or top of the webpage?  Or does the color picker component already have that logic?

Reply:
Currently, the color picker appears at a fixed Y position relative to the plot... which a quick test shows is a problem when the plot is near the bottom of the screen. I can take a pass at fixing that.

Reply:
I think we already have that logic for the channel colors.